### PR TITLE
[JENKINS-32623] Use real URL for user

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-32623](https://issues.jenkins-ci.org/browse/JENKINS-32623).

No tests because not worth it.

Untested so far, will give the PR build a try.

### Proposed changelog entries

* Bug: Fix link from build cause to user in case of unusual user names.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)

### CODE REVIEW

File: core/src/main/resources/hudson/model/Cause/UserIdCause/description.properties [hunk 1]
```diff
@@ -20,5 +20,5 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-started_by_user=Started by user <a href="{2}/user/{0}">{1}</a>
+started_by_user=Started by user <a href="{2}/{0}">{1}</a>
```
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/Cause/UserIdCause/description.properties]: Does it generate the correct number of slashes in URL?

File: core/src/main/java/hudson/model/Cause.java [hunk 2]
```diff
@@ -410,6 +412,11 @@ public String getUserName() {
             return userId == null ? "anonymous" : User.get(userId).getDisplayName();
         }
 
+        @Restricted(DoNotUse.class) // for Jelly
+        public String getUserUrl() {
+            return userId == null ? null : User.get(userId).getUrl();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Cause.java]: Should be http://javadoc.jenkins.io/hudson/model/User.html#get-java.lang.String-boolean-java.util.Map- instead. We do not want to occasionally re-create user instances in the case of race conditions (e.g. accessing page with `UserIdCause` in parallel with user deletion) 🐛

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
http://javadoc.jenkins.io/hudson/model/User.html#get-java.lang.String-boolean-java.util.Map- should be used in this case. Actually I would argue that `User.get(String)` should be deprecated at all since it causes much confusion

File: core/src/main/java/hudson/model/Cause.java (hunk 2 above)
Comment (daniel-beck) [reply core/src/main/java/hudson/model/Cause.java]: `getUserName` (5 lines above) does it already, so this PR doesn't make it worse.

File: core/src/main/resources/hudson/model/Cause/UserIdCause/description.properties (hunk 1 above)
Comment (daniel-beck) [reply core/src/main/resources/hudson/model/Cause/UserIdCause/description.properties]: It does not include a leading slash, so, yes.

**Review by oleg-nenashev — APPROVED** [review]
Screw these old `getUser()` APIs, I will burn them with fire at some point. 

If you are fine with fixing them in this class while you are, it would be nice. But I agree it's not mandatory.

**Comment by daniel-beck** [comment]
@oleg-nenashev Thanks. The only reason I bothered is that this was so straightforward, add a bunch of cleanup duty on top and I don't have the time 🙁

**Pull request merged**